### ISSUE TITLE
Some problems with loss

### ISSUE THREAD

**Issue by ydmo** [eriklindernoren/PyTorch-YOLOv3#108]
Hi~ I have some problems about the loss of classcification.

When I train the yolo, I found that loss will never become zero, although i use a single image as trainset and valset. So I found some problems about the calculation of loss:

In modules.py
self.ce_loss = nn.CrossEntropyLoss()  # Class loss
pred_cls = torch.sigmoid(prediction[..., 5:])  # Cls pred.
loss_cls = (1 / nB) * self.ce_loss(pred_cls[mask], torch.argmax(tcls[mask], 1))

Using sigmoid in pred_cls, may make the loss_cls never be optimized to zero!? Because nn.CrossEntropyLoss() combines nn.LogSoftmax() and nn.NLLLoss() in one single class.

Is that normal?

**Comment by fengxinjie** [comment]
hi, I think you are right, there is problem with  loss_cls, did you fix it?

**Comment by fengxinjie** [comment]
> Hi~ I have some problems about the loss of classcification.
> 
> When I train the yolo, I found that loss will never become zero, although i use a single image as trainset and valset. So I found some problems about the calculation of loss:
> 
> In modules.py
> self.ce_loss = nn.CrossEntropyLoss() # Class loss
> pred_cls = torch.sigmoid(prediction[..., 5:]) # Cls pred.
> loss_cls = (1 / nB) * self.ce_loss(pred_cls[mask], torch.argmax(tcls[mask], 1))
> 
> Using sigmoid in pred_cls, may make the loss_cls never be optimized to zero!? Because nn.CrossEntropyLoss() combines nn.LogSoftmax() and nn.NLLLoss() in one single class.
> 
> Is that normal?

change:
           pred_cls = torch.sigmoid(prediction[..., 5:])  # Cls pred.
to:
           pred_cls = prediction[..., 5:]  # Cls pred.


 sigmoid  is not need   for nn.CrossEntropyLoss()

**Comment by eriklindernoren** [comment]
There have been some recent changes to loss computation etc. Can you try with current master?
